### PR TITLE
models: invoice and invoice_item, factories, model specs, annotate

### PR DESCRIPTION
**Pull request by yunixon**

### CODE REVIEW

File: app/controllers/invoices_controller.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
+class InvoicesController < ApplicationController
+  layout 'settings'
```
Comment (shir) [inline app/controllers/invoices_controller.rb]: @yunixon don't use "settings" layout for this pages.

File: app/models/invoice.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+#  sent_at         :datetime
+#  paid_at         :datetime
+#  created_at      :datetime
+#  updated_at      :datetime
+#
+
+class Invoice < ActiveRecord::Base
+  monetize :amount_cents, with_model_currency: :currency
+
+  scope :ordered, -> { order('created_at DESC') }
```
Comment (shir) [inline app/models/invoice.rb]: @yunixon order in model:
1. constants
2. associations
3. Fields declaration (monetize, attachments, attr_accessor, accepts_nested_attriutes)
4. validations
5. scopes
6. class methods
7. normal methods

File: app/models/invoice.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+  belongs_to :organization, inverse_of: :invoices
+  belongs_to :customer, inverse_of: :invoices
+  has_many :invoice_items, inverse_of: :invoice, dependent: :destroy
+  accepts_nested_attributes_for :invoice_items,
+    reject_if: proc { |param| param[:amount].blank? },
```
Comment (shir) [inline app/models/invoice.rb]: @yunixon use `reject_if: :all_blank`

File: app/models/invoice.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    allow_destroy: true
+
+  validates :organization, :customer_id, :ends_at, :amount, :currency, presence: true
```
Comment (shir) [inline app/models/invoice.rb]: @yunixon one validation per attribute.
```
validates organization
validates customer
etc
```

File: app/models/invoice_item.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+class InvoiceItem < ActiveRecord::Base
+  monetize :amount_cents, with_model_currency: :currency
+
+  belongs_to :invoice, inverse_of: :invoice_items
+  belongs_to :customer
+
+  validates :invoice, :amount, presence: true
+  validates :amount, numericality: { greater_than: 0,
+    less_than_or_equal_to: Dictionaries.money_max }
+  validate :customer_or_description
```
Comment (shir) [inline app/models/invoice_item.rb]: @yunixon you may do it with native validation:
```rb
validates :description, presence: true, unless: :customer_id
```

File: app/views/home/_settings.html.slim [hunk 6]
```diff
@@ -13,3 +13,5 @@
       active: [['categories']]
     = active_link_to 'Members', members_path, class: 'list-group-item',
       active: [['members'],['index']]
+    = active_link_to 'Invoices', invoices_path, class: 'list-group-item',
```
Comment (shir) [inline app/views/home/_settings.html.slim]: @yunixon it's not setting

File: app/views/invoices/_form.html.slim [hunk 7]
```diff
@@ -0,0 +1,34 @@
+.row
+  .col-sm-6
+    = simple_form_for(@invoice, html: { class: 'invoice' }) do |f|
+      = f.error_notification
+
+      .form-inputs
+        = f.input :currency, collection: Dictionaries.currencies, autofocus: true
+        = f.input :customer_id, as: :select, placeholder: 'Customer',
```
Comment (shir) [inline app/views/invoices/_form.html.slim]: @yunixon use select2. Same way as in transaction

File: app/views/invoices/edit.html.slim [hunk 8]
```diff
@@ -0,0 +1,9 @@
+.page-header
+  h2 Edit invoice
+
+= render 'form'
+
+= link_to 'Back', invoices_path
+
+= content_for :sidebar
```
Comment (shir) [inline app/views/invoices/edit.html.slim]: @yunixon don't show settings sidebar.

File: app/views/invoices/_invoice_item_fields.html.slim [hunk 9]
```diff
@@ -0,0 +1,9 @@
+tr.nested-fields
+  td
+    = f.input :customer_id, as: :select, placeholder: 'Customer',
```
Comment (shir) [inline app/views/invoices/_invoice_item_fields.html.slim]: @yunixon use select2. Same way as in transaction

File: app/views/invoices/index.html.slim [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+        th Invoice total
+        th Sent date
+        th Paid date
+    tbody
+      = render @invoices
+- else
+  .alert.alert-warning No invoices
+
+#paginator.col-md-offset-5
+  = paginate @invoices, remote: true
```
Comment (shir) [inline app/views/invoices/index.html.slim]: @yunixon remove `remote: true`

File: app/views/invoices/index.js.erb [hunk 11]
```diff
@@ -0,0 +1,2 @@
+$('.invoices').html('<%= j render(@invoices) %>');
```
Comment (shir) [inline app/views/invoices/index.js.erb]: @yunixon remove this view.

File: app/views/invoices/show.html.slim [hunk 12]
```diff
@@ -0,0 +1,37 @@
+.page-header
+  .pull-right
+    = link_to 'Back', invoices_path, class: 'btn btn-primary'
+  h1 = "Invoice ##{@invoice.id}"
```
Comment (shir) [inline app/views/invoices/show.html.slim]: @yunixon for title use "Invoice #{customer} from #{date}"

File: app/models/invoice.rb (hunk 2 above)
Comment (yunixon) [inline app/models/invoice.rb]: @shir Thanks! I've only seen this:
https://github.com/bbatsov/rails-style-guide/blob/master/README.md#macro-style-methods

File: app/assets/javascripts/invoices.js [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+    }
+  });
+
+  var name = $customerField.data('value');
+  if (name) {
+    $customerField.select2("data", { id: name, text: name });
+  }
+}
+
+function addInvoiceNestedCustomerSelect2() {
```
Comment (shir) [inline app/assets/javascripts/invoices.js]: @yunixon why do you create same function for different fields? Why can't you one function for all customer fields?

File: app/helpers/invoices_helper.rb [hunk 14]
```diff
@@ -0,0 +1,9 @@
+module InvoicesHelper
+  def invoice_date_range(invoice)
+    if invoice.starts_at
```
Comment (shir) [inline app/helpers/invoices_helper.rb]: @yunixon
```rb
[starts_at, ends_at].compact.map{ |d| l d }.join(' - ')
```

File: app/models/invoice_item.rb [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+  validates :invoice, presence: true
+  validates :amount, presence: true
+  validates :description, presence: true, unless: :customer_id?
+  validates :amount, numericality: { greater_than: 0,
+    less_than_or_equal_to: Dictionaries.money_max }
+
+  before_validation :find_customer, if: Proc.new{ self.customer_name.present? }
+
+  def customer_name=(value)
```
Comment (shir) [inline app/models/invoice_item.rb]: @yunixon put all this customer search stuff it in concern and use in all models

File: app/assets/javascripts/application.js [hunk 16]
```diff
@@ -34,3 +36,53 @@ function scrollTo($el) {
 $(function () {
   $('[data-toggle="tooltip"]').tooltip();
 })
+
+$(function () {
```
Comment (Gataniel) [inline app/assets/javascripts/application.js]: Why don't we put it in above 
```
$(function () {
   $('[data-toggle="tooltip"]').tooltip();
   addCustomerSelect2('.customer-select2');
```

File: app/assets/javascripts/application.js [hunk 17]
```diff
@@ -34,3 +36,53 @@ function scrollTo($el) {
...
+  addCustomerSelect2('.customer-select2');
```
Comment (Gataniel) [inline app/assets/javascripts/application.js]: Seems like we can send multiple fields to `function addCustomerSelect2(fields)` hence do it in one line

File: app/assets/javascripts/application.js [hunk 18]
```diff
@@ -34,3 +36,53 @@ function scrollTo($el) {
...
+  addCustomerSelect2('.nested-customer-select2');
```
Comment (Gataniel) [inline app/assets/javascripts/application.js]: Seems like we can send multiple fields to `function addCustomerSelect2(fields)` hence do it in one line

File: app/models/invoice.rb [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+# Table name: invoices
+#
+#  id              :integer          not null, primary key
+#  organization_id :integer          not null
+#  customer_id     :integer          not null
+#  starts_at       :datetime
+#  ends_at         :datetime         not null
+#  currency        :string           default("USD"), not null
+#  amount_cents    :integer          default(0), not null
+#  amount_currency :string           default("USD"), not null
```
Comment (AlexanderUlitin) [inline app/models/invoice.rb]: @yunixon please clarify why we need 2 currency fields for this model?

File: app/views/invoices/_form.html.slim [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+      input_html: { class: 'datepicker' }
+    #invoice
+      h4 Invoice items
+      table.table.invoice-form-items
+        thead
+          tr
+            th
+            th
+            th
+            th
```
Comment (AlexanderUlitin) [inline app/views/invoices/_form.html.slim]: @yunixon remove this empty row, or add some column names.

File: app/views/invoices/_invoice_item_fields.html.slim [hunk 21]
```diff
@@ -0,0 +1,10 @@
+tr.nested-fields
+  td
+    = f.input :customer_name, input_html: { class: 'nested-customer-select2',
+      data: { url: autocomplete_customers_path, placeholder: 'Customer',
+      value: current_organization.find_customer_name_by_id(f.object.customer_id) } }
+    = f.input :amount, input_html: { class: 'nested-amount' }
+  td
+    = f.input :hours, input_html: { class: 'nested-hours' }
+    = f.input :description, input_html: { class: 'nested-description' }
```
Comment (AlexanderUlitin) [inline app/views/invoices/_invoice_item_fields.html.slim]: @yunixon will it be discplayed like textarea or like single-line textfield?

Comment (yunixon) [inline app/views/invoices/_invoice_item_fields.html.slim]: @AlexanderUlitin like textarea

File: spec/features/create_invoice_spec.rb [hunk 22]
```diff
@@ -0,0 +1,42 @@
+require 'spec_helper'
```
Comment (AlexanderUlitin) [inline spec/features/create_invoice_spec.rb]: @yunixon please move all this specs to `spec/features/invoices/` directory and name like `create_spec.rb`, `index_spec.rb` e.t.c

File: spec/features/create_invoice_spec.rb [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+  context 'Create invoice without items' do
+    before do
+      click_on 'New Invoice'
+      select2 customer.name, css: '#s2id_invoice_customer_name', search: true
+      fill_in 'Ends at', with: Time.now.strftime('%d/%m/%Y')
+      fill_in 'invoice[amount]', with: amount_str
+      click_on 'Create Invoice'
+    end
+
+    it { expect(page).to have_content 'Invoice was successfully created' }
```
Comment (AlexanderUlitin) [inline spec/features/create_invoice_spec.rb]: @yunixon add more check after creating: check if there are new invoice on the page with link to view invoice.

File: spec/features/create_invoice_spec.rb [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+    before do
+      click_on 'New Invoice'
+      select2 customer.name, css: '#s2id_invoice_customer_name', search: true
+      fill_in 'Ends at', with: Time.now.strftime('%d/%m/%Y')
+      fill_in 'invoice[amount]', with: amount_str
+      click_on 'Add item'
+      find('#invoice .nested-fields input.nested-amount').set(amount_str)
+      find('#invoice .nested-fields input.nested-hours').set('1.1')
+      find('#invoice .nested-fields textarea.nested-description').set('Description')
+      click_on 'Create Invoice'
```
Comment (AlexanderUlitin) [inline spec/features/create_invoice_spec.rb]: @yunixon add one move InvoiceItem in this case.

File: spec/features/create_invoice_spec.rb [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+    end
+
+    it { expect(page).to have_content 'Invoice was successfully created' }
```
Comment (AlexanderUlitin) [inline spec/features/create_invoice_spec.rb]: @yunixon add more check after creating: check if there are new invoice on the page with link to view invoice.

File: spec/features/invoices_list_spec.rb [hunk 26]
```diff
@@ -0,0 +1,19 @@
+require 'spec_helper'
```
Comment (AlexanderUlitin) [inline spec/features/invoices_list_spec.rb]: @yunixon rename this spec to index_spec.rb

File: spec/features/invoices_page_spec.rb [hunk 27]
```diff
@@ -0,0 +1,24 @@
+require 'spec_helper'
```
Comment (AlexanderUlitin) [inline spec/features/invoices_page_spec.rb]: @yunixon combine this spec with previous one.

File: spec/features/create_invoice_spec.rb (hunk 22 above)
Comment (AlexanderUlitin) [inline spec/features/create_invoice_spec.rb]: @yunixon also add a basic spec for Edit and Update Invoice and for Delete Invoice.

File: app/views/invoices/_invoice_item_fields.html.slim (hunk 21 above)
Comment (AlexanderUlitin) [inline app/views/invoices/_invoice_item_fields.html.slim]: @yunixon cool, thanks.

**Pull request merged**